plugins/outputs/websocket/websocket.go

```diff
@@ -0,0 +1,216 @@
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+
+  ## Data format to output.
+  ## Each data format has it's own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md
+  # data_format = "influx"
+
+  ## Additional HTTP Upgrade headers
+  # [outputs.websocket.headers]
+  #   Authorization = "Bearer <TOKEN>"
+`
+
+const (
+	defaultConnectTimeout = 30 * time.Second
+	defaultWriteTimeout   = 30 * time.Second
+	defaultReadTimeout    = 30 * time.Second
+)
+
+// WebSocket can output to WebSocket endpoint.
+type WebSocket struct {
+	URL            string            `toml:"url"`
+	ConnectTimeout config.Duration   `toml:"connect_timeout"`
+	WriteTimeout   config.Duration   `toml:"write_timeout"`
+	ReadTimeout    config.Duration   `toml:"read_timeout"`
+	Headers        map[string]string `toml:"headers"`
+	UseTextFrames  bool              `toml:"use_text_frames"`
+	Logger         telegraf.Logger   `toml:"-"`
+	tls.ClientConfig
+
+	conn       *ws.Conn
+	serializer serializers.Serializer
+}
+
+// SetSerializer implements serializers.SerializerOutput.
+func (w *WebSocket) SetSerializer(serializer serializers.Serializer) {
+	w.serializer = serializer
+}
+
+// Description of plugin.
+func (w *WebSocket) Description() string {
+	return "Generic WebSocket output writer."
+}
+
+// SampleConfig returns plugin config sample.
+func (w *WebSocket) SampleConfig() string {
+	return sampleConfig
+}
+
+var errInvalidURL = errors.New("invalid websocket URL")
+
+// Init the output plugin.
+func (w *WebSocket) Init() error {
+	if parsedURL, err := url.Parse(w.URL); err != nil || (parsedURL.Scheme != "ws" && parsedURL.Scheme != "wss") {
+		return fmt.Errorf("%w: \"%s\"", errInvalidURL, w.URL)
+	}
+	return nil
+}
+
+// Connect to the output endpoint.
+func (w *WebSocket) Connect() error {
+	tlsCfg, err := w.ClientConfig.TLSConfig()
+	if err != nil {
+		return fmt.Errorf("error creating TLS config: %v", err)
+	}
+
+	dialer := &ws.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: time.Duration(w.ConnectTimeout),
+		TLSClientConfig:  tlsCfg,
+	}
+
+	headers := http.Header{}
+	for k, v := range w.Headers {
+		headers.Set(k, v)
+	}
+
+	conn, resp, err := dialer.Dial(w.URL, headers)
+	if err != nil {
+		return fmt.Errorf("error dial: %v", err)
+	}
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusSwitchingProtocols {
+		return fmt.Errorf("wrong status code while connecting to server: %d", resp.StatusCode)
+	}
+
+	w.conn = conn
+	go w.read(conn)
+
+	return nil
+}
+
+func (w *WebSocket) read(conn *ws.Conn) {
+	defer func() { _ = conn.Close() }()
+	if w.ReadTimeout > 0 {
+		if err := conn.SetReadDeadline(time.Now().Add(time.Duration(w.ReadTimeout))); err != nil {
+			return
```

Comment:
Please log the error here so we can trace down the failure if necessary.

Reply:
Done